### ISSUE TITLE
Transfer Rates wrong and Transfer Sizes go negative

### ISSUE THREAD

**Issue by JochenB** [syncthing/syncthing-gtk#55]
Just spotted while watching a transfer...

The Upload/Download Rate shown in Syncthing GTK/Linux is totally bogus. I see values of only a few kBps ("random" numbers between 7 through 57) while Syncthing GTK on Windows shows me a average of almost about 20MBps (and that's a whole lot closer to the increase in sync percentage at each update in relation to the overall size). Also, the associated amount of data it shows next to the transfer speed is totally bogus. Once it reaches 2GB it reverts to negative figures (i.e. the current sync job just stopped at "0 Bps (-255777153 B)") goes up to about negative 8 billion before it reverts to zero - int32 overflow?). However, seems it's also not really correct on Windows as it shows only "19.13 MBps (1.83 GB)" downloaded while it was actually 7.76GB worth of data. Having added another 10GB of data just now ... now the box (in Windows, receiving side) remains frozen (top/left - doesn't update speeds nor sizes).

What strikes me as well ... why is the box of the other peer not changing? On the sending side (where the data was dropped into the "repo") it updates the percentage in the top/right box, but the box of the peer to where it sync steadily shows rate/size at "0 bps (0 B)". On the receiving side it's about the same. The top/right box updates data (though that doesn't seem to work reliable - see paragraph above), but the rate/size in the box of the peer that is sending stays steadily at "0 bps (0 B)".

Unless I miss something here, the box(es) of the peer(s) to where a sync is going (sending side) as well the box(es) of the peer(s) from where a sync is coming from (receiving side) should update the rate and size (that's actually what I am fighting with at the latest revision of my C# port - I have a rough prototype up and running that needs a lot of refinement until it's "production usage" ready, currently dealing with filling stats into the UI).

The figures shown inside the "repo" boxes on the left is correct ... it reflects the correct amount of data and correct count of files (though I haven't tried to see if it would also go negative at some file count).

Edit: the "added yet another 10GB" sync finished ... now STGTK/Linux shows "0 bps (-1567673409 B)" at the "Upload Rate" in the top/right box. The figures in the "repo" box is correct ... 18.48GB and 41 files". STGTK/Windows didn't update the figures in the top/right box, they stayed "frozen" at the last values displayed (Download Rate 19.13 MBps (1.83GB) / Upload Rate 19.13 MBps (2.04MB). The "repo" is showing correct figures though. /Edit

**Comment by kozec** [comment]
Ok, I pretty-much copied code from [syncthingController.js]{https://github.com/syncthing/syncthing/blob/34cb3057556ca532f4064710af5cdc4b6e575b2e/gui/scripts/syncthing/core/controllers/syncthingController.js#L323} source and ST-GTK gives roundly same rates as WebUI does.

It's probably impossible to get identical values at any given time, as refresh rates of WebUI and ST-GTK are different and sampled at different times, but that shouldn't be a big problem. Thanks to delays in HTTP response is neither value 100% accurate anyway :)

**Comment by JochenB** [comment]
I wouldn't worry about the transfer rate values being non-identical between the web ui and the app - as you said they are rough estimates anyway, and polling more frequently only increases overall load (to draw a somewhat related comparison: Torrent clients usually poll / refresh the UI in 1000ms intervals by default).

Speaking about: A more direct API to Syncthing would be a nice thing to have... having to wrap around the HTTP interface is not a very elegant solution.

However, one thing that popped into mind while tinkering around in C# (should be doable in Python/GTK as well) ...
If we already wrap around the HTTP interface, why don't we simply do it losely like some other closed-source P2P synchronizer, who shall remain unnamed, does: create a customized "asset" for the WebUI (to trim it down to some sane-sized layout that would fit into a window) and render that in an app window through an integrated render engine (i.e. webkit, gecko - to not depend on the horror that is IE/Trident). Starting/Stopping the daemon remains the same ... and ... auto-update plus daemon config also remains the same ... the only thing that effectively changes is that the UI is, more-or-less, passed-through through a custom asset, and that way it would even be possible to setup a "Monitor other instances" on remote systems to where the user has access. I'd like to give that idea a try.

**Comment by kozec** [comment]
@JochenB 

> some other closed-source P2P synchronizer, who shall remain unnamed

Unnamed and thus unrecognized :D Sorry, I only used Dropbox before Syncthing.

I see three problems with "customized webapp in browser frame" approach:

1st, it's much more resource-heavy. Browser frame will end taking CPU and memory for application _plus_ original WebUI.
2nd, it prevents UI from accessing daemon that is not made ready for it. Just starting Syncthing as rc.d service would break UI completely.
And third, it's very, very limited. Even "browse directory" menu would be problematic (although probably not impossible) from javascript...

**Comment by AudriusButkevicius** [comment]
@JochenB I have a proof of concept of that in C# using CefSharp in case you are interested messing around with it.
Never finished it though.

**Comment by JochenB** [comment]
@kozec 

> Unnamed and thus unrecognized :D Sorry, I only used Dropbox before Syncthing.

The software I mean is "BTSync". With their v1.4 trainwreck (which drove a good amount of users over to Syncthing) they render the interface through IE/Trident (therefore making IE8+ mandatory on Windows) so they don't need to employ manly coding monkeys (though they formerly had a very nice C++ control interface they simply did away with).

> I see three problems with "customized webapp in browser frame" approach:
> 1st, it's much more resource-heavy. Browser frame will end taking CPU and memory for application plus original WebUI.

Not really. It only increases the installation size due to having to ship the internal render engine (not much of a problem with Linux where gecko/xulrunner is part of a Cinnamon/Gnome3/MATE/Unity install anyway) - though, let's be honest ... who cares about a couple of MB at the days and and age of Terabyte drives? As for being more resource heavy: It's just the render engine (the very core part that renders HTML/CSS and executes JavaScript/AJAX), doesn't really hog resources as there's no full-blown browser framework glued around it. That's orders of magnitude below the usage and resource.-hogginess of Firefox or Chrome (we're talking a few MB RAM usage here, not several hundred of MB).

> 2nd, it prevents UI from accessing daemon that is not made ready for it. Just starting Syncthing as rc.d service would break UI completely.

That sounds more like a Linux specific headache than a cross-platform headache (syncthing doesn't come as a Service on Windows, you need "messing about with Windows" skills or third party tools to strap syncthing in as a Service.

> And third, it's very, very limited. Even "browse directory" menu would be problematic (although probably not impossible) from javascript...

"Browse Directory"? You got somethin' cookin' there? Unless I miss out on something there's no "Open repo directory" button in STGTK yet. ;)

However, firing up explorer.exe (on Windows, obviously) with the corresponding repo directory is a piece of cake... just a slightly unholy mess because Java(Script).

@AudriusButkevicius 

Well, I love to look through pieces of code to maybe find the one or another inspiration or "better practice" idea ... at the moment I'm just toying around in Visual Studio with a "bare skeleton concept" to a) try to replicate STGTK in C# so kozec can be freed from maintaining "the_hell" ;) and b) throwing around some ideas to maybe cough up a better UI (like "why not do it the 'BTSync' way and simply render the UI through custom assets and some regex magic oneself").

If you want to share the code ZIP the VS project up and put it on some file host.

Mono would actually be the way to go if kozec would know C#, but I dare to bet the Mono idea will most likely only get me "yea, let's depend on 100MB+ of patent-encumbered dependecies on Linux" complaint. ;) *ducks*

Anyhow, I'll pursue the "embed the HTML UI" idea for a few days to see how problematic it would really be (by keeping the "Browse repo" idea in mind).

**Comment by kozec** [comment]
@JochenB 

> "Browse Directory"? You got somethin' cookin' there? Unless I miss out on something there's no "Open repo directory" button in STGTK yet. ;)

True, there is no such button, but it's last option in context menu. Just right-click on folder and select `Browse` ;-)
But yeah, I guess there is some method how to call native method from javascript running in browser frame, so it may be not as big problem as I imagine.

By the way, after porting is done, it's not that big problem to maintain windows port. And I had "THE_HELL" variable <A href="https://github.com/syncthing/syncthing-gtk/blob/c3e21560a07947c095552e9009027ce262463fa6/syncthing_gtk/statusicon.py#L14">resreved for Ubuntu actually</A> :D

Using Mono + WindowsForm on Linux Windows is pretty-much same mess as using GTK on Windows, so although I know some of C#, it's not way I would go for multiform app.

**Comment by JochenB** [comment]
> True, there is no such button, but it's last option in context menu. Just right-click on folder and select Browse ;-)

Whoa... I totally missed out on that. X-|

> By the way, after porting is done, it's not that big problem to maintain windows port. And I had "THE_HELL" variable resreved for Ubuntu actually :D

I was only browsing through your sources on the surface... seems I understood that "if" wrong (thought it would refer to Windows). Though... Ubuntu isn't as bad... once you use Linux Mint (Cinnamon) all of that "UI Hell" headache goes away.

**Comment by AudriusButkevicius** [comment]
@kozec most browser components (even the shitty WinForms built-in one which relies on IE) have some way of injecting a javascript object to the browser window which when accessed bubbles through to native code (C# in my case).

Further, CefSharp has a way to manipulate the Dom, so what I did, was wrote my own angular controller which I then at load time replaced syncthings controller with which then called methods on this "injected" object which as a result triggered stuff in the native code.

I did it a while ago, I'll put it up on github for inspiration for hackery.
